File: 1096-brace-expansion-ii/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    // Entry point
    vector<string> braceExpansionII(string expression) {
        int pos = 0;
        unordered_set<string> result = parseConcat(expression, pos);
        vector<string> ans(result.begin(), result.end());
        sort(ans.begin(), ans.end());
        return ans;
    }

private:
    // Parse a concatenation sequence until ',' or '}'
    unordered_set<string> parseConcat(const string& s, int& i) {
        unordered_set<string> cur = {""};          // start with empty string
        while (i < (int)s.size() && s[i] != '}' && s[i] != ',') {
            unordered_set<string> next;
            if (s[i] == '{') {
                ++i;                                 // skip '{'
                next = parseSet(s, i);               // parse the union inside braces
                ++i;                                 // skip '}'
            } else {                                 // a single letter
                next.insert(string(1, s[i]));
                ++i;
            }
            cur = cartesianProduct(cur, next);
        }
        return cur;
    }

    // Parse a union set: { expr1, expr2, ... }
    unordered_set<string> parseSet(const string& s, int& i) {
        unordered_set<string> total;
        while (true) {
            unordered_set<string> part = parseConcat(s, i);
            total.insert(part.begin(), part.end()); // union
            if (i >= (int)s.size() || s[i] == '}')
                break;
            // s[i] must be ',' here
            ++i; // skip ','
        }
        return total;
    }

    // Cartesian product of two sets of strings
    unordered_set<string> cartesianProduct(const unordered_set<string>& a,
                                            const unordered_set<string>& b) {
        unordered_set<string> res;
        for (const string& x : a) {
            for (const string& y : b) {
                res.insert(x + y);
            }
        }
        return res;
    }
};

File: 1096-brace-expansion-ii/solution.py
from typing import List, Set

class Solution:
    def braceExpansionII(self, expression: str) -> List[str]:
        """
        Evaluate the expression according to the grammar and return a sorted list
        of distinct words.
        """
        self.s = expression
        self.n = len(expression)
        self.idx = 0
        result = self._parse_concat()
        return sorted(result)

    def _parse_concat(self) -> Set[str]:
        """Parse a concatenation sequence until ',' or '}'."""
        cur = {""}                     # start with empty string
        while self.idx < self.n and self.s[self.idx] not in ('}', ','):
            if self.s[self.idx] == '{':
                self.idx += 1          # skip '{'
                nxt = self._parse_set()
                self.idx += 1          # skip '}'
            else:
                nxt = {self.s[self.idx]}
                self.idx += 1
            cur = {a + b for a in cur for b in nxt}
        return cur

    def _parse_set(self) -> Set[str]:
        """Parse a union set: { expr1, expr2, ... }."""
        total = set()
        while True:
            part = self._parse_concat()
            total.update(part)          # union
            if self.idx >= self.n or self.s[self.idx] == '}':
                break
            # current char must be ','
            self.idx += 1               # skip ','
        return total